two_factor_otp_auth/models/res_users.py

```diff
@@ -0,0 +1,175 @@
+            "secret_code_2fa": False,
+        }
+        self.write(values)
+
+    def action_disable_2f_auth(self):
+        """
+        Set `enable_2fa` field value to `False`. Check access for action
+        via `_can_change_2f_auth_settings`.
+        """
+        self._can_change_2f_auth_settings()
+        values = {
+            "enable_2fa": False,
+        }
+        self.write(values)
+
+    def action_enable_2f_auth(self):
+        """
+        Set `enable_2fa` field value to `False`. Check access for action
+        via `_can_change_2f_auth_settings`.
+        """
+        self._can_change_2f_auth_settings()
+        values = {
+            "enable_2fa": True,
+        }
+        self.write(values)
+
+    def _check_credentials(self, password):
+        """
+        Overload core method to also check Two Factor Authentication
+        credentials.
+
+        Raises:
+         * odoo.addons.two_factor_otp_auth.exceptions.MissingOtpError - no
+            `otp_code` in request params. Should be caught by controller and
+            render and open enter "one-time-password" page or QR code creation
+        """
+        self.ensure_one()
+        super(ResUsers, self)._check_credentials(password)
+        if self.enable_2fa:
+            params = request.params
+            secret_code = self.secret_code_2fa or params.get("secret_code_2fa")
+            if params.get("otp_code") is None:
+                request.session.otk_uid = self.id
+                raise MissingOtpError()
+            else:
+                # can trigger `InvalidOtpError`
+                self._check_otp_code(
+                    params.pop("otp_code"),
+                    secret_code,
+                )
+
+    def _generate_secrets(self):
+        """
+        Generate QR-Code based on random set of letters
+
+        Returns:
+         * tuple - generated secret_code and binary qr-code
+        """
+        self.ensure_one()
+        key = b32encode(urandom(10))
+        code = pyotp.totp.TOTP(key).provisioning_uri(self.login)
+        img = qrcode.make(code)
+        _, file_path = mkstemp()  # creating temporary file
+        img.save(file_path)
+
+        with open(file_path, 'rb') as image_file:
+            qr_image_code = b64encode(image_file.read())
+
+        remove(file_path)  # removing temporary file
+        return key, qr_image_code
+
+    @api.model
+    def _can_change_2f_auth_settings(self):
+        """
+        Check that current user can make mass actions with Two Factor
+        Authentication settings.
+
+        Raises:
+         * odoo.exceptions.AccessError: only administrators can do this
+           action
+
+        TODO:
+         * Rewrite text of warning - add list of groups with access
+         * Or even better create separate group.
+        """
+        if not self.env.user._is_admin():
+            raise AccessError(_(
+                "Only Administrators can do this operation!"
+            ))
```

Comment:
Raise exception if user do not belong to the group "Technical Settings".

Reply:
done